[2.7.2] Update PSS docs (#479)
rancher/rancher-docs: docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/pod-security-standards.md

title: Pod Security Standards
---

[Pod Security Standards (PSS)](https://kubernetes.io/docs/concepts/security/pod-security-standards/) define security restrictions for a broad set of workloads. They replaced, the now removed, [Pod Security Policies](https://kubernetes.io/docs/concepts/security/pod-security-policy/) in Kubernetes v1.25 and above.

## Upgrade to Pod Security Standards

> Make sure that all Pod Security Policies have been migrated, which should result in equivalent security controls still being enforced (e.g. [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/), [Gatekeeper](https://github.com/open-policy-agent/gatekeeper), [Kubewarden](https://www.kubewarden.io/), [Kyverno](https://kyverno.io/), [NeuVector](https://neuvector.com/)) and all PodSecurityPolicy objects removed from the cluster.

Reviewer: ```suggestion
Make sure that you migrate all Pod Security Policies. This should grant equivalent security controls through [Pod Security Admissions](https://kubernetes.io/docs/concepts/security/pod-security-admission/), or through solutions such as [Gatekeeper](https://github.com/open-policy-agent/gatekeeper), [Kubewarden](https://www.kubewarden.io/), [Kyverno](https://kyverno.io/), and [NeuVector](https://neuvector.com/). Remove all PodSecurityPolicy objects from the cluster.
```

:::info important

> The new policy enforcement must be added before the PodSecurityPolicy objects are removed, or you risk creating a window of opportunity for privilege escalation within the cluster.

Reviewer: ```suggestion
You must add your new policy enforcement mechanisms _before_ you remove the PodSecurityPolicy objects. If you don't, you may create an opportunity for privilege escalation attacks within the cluster.
```

:::

## Pod Security Admission Configuration Templates

> Rancher comes with two Pod Security Admission Configuration (PSAC) templates that you can assign to a cluster:

Reviewer: In some places internally we use PSACT, but technically it's `kind: PodSecurityAdmissionConfiguration` when users go to create one. Imho I'd prefer something like _Pod Security Admission (PSA) configuration template_ or _PSA configuration template_ on first mention, then _template_ for short.

> - `rancher-privileged`: This is the most permissive configuration and does not restrict the behavior of any pod, allowing for known privilege escalations. This policy contains no exemptions.

Reviewer: ```suggestion
- `rancher-privileged`: This is the most permissive configuration and doesn't restrict the behavior of the pod, allowing for known privilege escalations. This policy has no exemptions.
```

> - `rancher-restricted`: This policy defines a heavily restricted configuration, based on current pod hardening best practices. This policy contains namespace level exemptions for Rancher components, as described below.

Reviewer: ```suggestion
- `rancher-restricted`: This configuration is heavily restricted and follows current best practices for hardening pods. There are namespace-level exemptions for Rancher components, as described below.
```

If you are a Rancher administrator or have restricted administrator privileges, you can customize restrictions and permissions by creating additional PSAC templates, or by editing existing templates.

:::caution
If you edit an existing PSAC template while it is still in use, it *will* affect all clusters that have been assigned that template.
:::

If you want to allow users other than the Rancher administrator to manage PSAC templates, you can bind the user to a role that grants the user all verbs (`"*"`) on `management.cattle.io/podsecurityadmissionconfigurationtemplates`.

:::warning
Any user that is bound to the above permission will be able to change the restriction levels on all managed clusters which use this template, including ones they have no permissions on.
:::

### Rancher on PSA-restricted Clusters

> When you run Rancher on a Kubernetes cluster that enforces a restrictive PSS policy by default, such as the `rancher-restricted` PSS policy, you will need to exempt the following namespaces, otherwise the policy might prevent Rancher system pods from properly running.

Reviewer: There's some slippage here between PSA and PSS, as the heading uses "PSA-restricted" but the body text talks of PSS policies. They're closely related concepts with similar acronyms, and a reader might get confused on which is which. Maybe we can avoid the acronyms here by using "security policy"? Example: "[...] a restrictive security policy by default, such as the `rancher-restricted` policy".

- `ingress-nginx`
- `kube-system`
- `cattle-system`
- `cattle-epinio-system`
- `cattle-fleet-system`
- `longhorn-system`
- `cattle-neuvector-system`
- `cattle-monitoring-system`
- `rancher-alerting-drivers`
- `cis-operator-system`
- `cattle-csp-adapter-system`
- `cattle-externalip-system`
- `cattle-gatekeeper-system`
- `istio-system`
- `cattle-istio-system`
- `cattle-logging-system`
- `cattle-windows-gmsa-system`
- `cattle-sriov-system`
- `cattle-ui-plugin-system`
- `tigera-operator`

> The listed namespaces are already exempt in the Rancher provided `rancher-restricted` PSS policy. A sample Admission Configuration with all the required exemptions to run Rancher can be found [here](psa-restricted-exemptions.yaml).

Reviewer: ```suggestion
The listed namespaces are already exempt in the built-in Rancher `rancher-restricted` security policy. Refer to this [sample Admission Configuration](psa-restricted-exemptions.yaml), which has all the exemptions you need to run Rancher.
```

:::

### Rancher On PSA-restricted Clusters
### Rancher on PSA-restricted Clusters

When you run Rancher on a Kubernetes cluster that enforces the `rancher-restricted` PSS policy by default, you'll need to exempt the following namespaces. This prevents the policy from blocking Rancher system pods:
When you run Rancher on a Kubernetes cluster that enforces a restrictive PSS policy by default, such as the `rancher-restricted` PSS policy, you will need to exempt the following namespaces, otherwise the policy might prevent Rancher system pods from properly running.

Here is the list of the namespaces that require an exemption:

- `ingress-nginx`
- `kube-system`
- `kube-public`
- `cattle-system`
- `cattle-alerting`
- `cattle-logging`
- `cattle-prometheus`
- `ingress-nginx`
- `cattle-global-data`
- `cattle-istio`
- `kube-node-lease`
- `cert-manager`
- `cattle-global-nt`
- `security-scan`
- `cattle-epinio-system`
- `cattle-fleet-system`
- `cattle-fleet-local-system`
- `calico-system`
- `tigera-operator`
- `cattle-impersonation-system`

> - `rancher-operator-system`

Reviewer 2: Can you clarify why all of these were removed ? Would like @crobby to confirm changes as well before we merge.

Author: Removed because they are not in the list of ns currently exempt in the policy that we provide in Rancher - https://github.com/rancher/rancher/blob/release/v2.7/pkg/data/management/podadmissionconfigurationtemplate_data.go#L12-L33.

CC @crobby

Reviewer 2: Ahh, that is because there are two lists. The template is for all downstream clusters and includes things like feature charts namespaces. There is a second list around system namespaces which is meant for local cluster: https://github.com/rancher/rancher/blob/release/v2.7/pkg/settings/setting.go#L27
Something like "cattle-impersonation-system" would only ever be on local.

Reviewer 2: Though @crobby can you remove `rancher-operator-system`, the operator has been removed completely in 2.7.

Reviewer 3: Yes, that all makes sense to me.

Author: @crobby and @cbron if you are fine with the updates, can you please approve the PR?

Author: @cbron and @crobby I updated the namespace exemption list, based on our offline discussion, through a separate commit to make it easier for review. PTAL.

> The listed namesapces are already exempt in the Rancher provided `rancher-restricted` PSS policy. A sample AdmissionConfiguration with all the required exemptions to run Rancher can be found [here](psa-restricted-exemptions.yaml).

Reviewer 4: ```suggestion
The listed namespaces are already exempt in the Rancher provided `rancher-restricted` PSS policy. A sample Admission Configuration with all the required exemptions to run Rancher can be found [here](psa-restricted-exemptions.yaml).
```

title: Pod Security Standards
---

> [Pod Security Standards (PSS)](https://kubernetes.io/docs/concepts/security/pod-security-standards/) define security restrictions for a broad set of workloads. They replace [Pod Security Policies](https://kubernetes.io/docs/concepts/security/pod-security-policy/) in Kubernetes v1.25 and above.

Reviewer: ```suggestion
[Pod Security Standards (PSS)](https://kubernetes.io/docs/concepts/security/pod-security-standards/) and [Pod Security admissions (PSA)](https://kubernetes.io/docs/concepts/security/pod-security-admission/) define security restrictions for a broad set of workloads. They replace [Pod Security Policies](https://kubernetes.io/docs/concepts/security/pod-security-policy/) in Kubernetes v1.25 and above.

PSS define security levels for workloads.  PSAs describe requirements for pod security contexts and related fields. PSAs reference PSS levels to define security restrictions.
```

I had a look at https://github.com/rancher/rancher-docs/issues/462, started writing, and realized that the issue overlaps with content in this PR. On this page, we move quickly from PSS to discussing PSA config templates, and on reflection, we need a little more of an explanation of how PSS and PSA relate, so that the section on templates feels more connected.

Reviewer: In addition -- this screenshot of the GUI: 
![image](https://user-images.githubusercontent.com/19174201/224775502-1c5d0a03-061c-4fe1-9309-eef7d57c6ea0.png)

The warning in the screenshot states that PSA is the replacement for PSPs, and doesn't mention PSS. Maybe this page should be retitled to Pod Security Admissions & Pod Security Standards, so that it's in sync with the GUI?

:::

### Rancher on PSA-restricted Clusters

When you run Rancher on a Kubernetes cluster that enforces a restrictive security policy by default, you will need to exempt the following namespaces, otherwise the policy might prevent Rancher system pods from properly running.

- `calico-apiserver`
- `calico-system`
- `cattle-alerting`
- `cattle-csp-adapter-system`
- `cattle-epinio-system`
- `cattle-externalip-system`
- `cattle-fleet-local-system`
- `cattle-fleet-system`
- `cattle-gatekeeper-system`
- `cattle-global-data`
- `cattle-global-nt`
- `cattle-impersonation-system`
- `cattle-istio`
- `cattle-istio-system`
- `cattle-logging`
- `cattle-logging-system`
- `cattle-monitoring-system`
- `cattle-neuvector-system`
- `cattle-prometheus`
- `cattle-sriov-system`
- `cattle-system`
- `cattle-ui-plugin-system`
- `cattle-windows-gmsa-system`
- `cert-manager`
- `cis-operator-system`
- `ingress-nginx`
- `istio-system`
- `kube-node-lease`
- `kube-public`
- `kube-system`
- `longhorn-system`
- `rancher-alerting-drivers`

> - `rancher-operator-system`

Reviewer 2: We don't need this anymore: 
```suggestion
```

- `security-scan`
- `tigera-operator`

> These namespaces are used by Rancher, some Rancher owned charts, and RKE2 and K3s distributions. A subset of the listed namespaces are already exempt in the built-in Rancher `rancher-restricted` policy for use in downstream clusters. For a complete template which has all the exemptions you need to run Rancher, please refer to this [sample Admission Configuration](psa-restricted-exemptions.yaml).

Reviewer: ```suggestion
Rancher, some Rancher owned charts, and RKE2 and K3s distributions all use these namespaces. A subset of the listed namespaces are already exempt in the built-in Rancher `rancher-restricted` policy, for use in downstream clusters. For a complete template which has all the exemptions you need to run Rancher, please refer to this [sample Admission Configuration](psa-restricted-exemptions.yaml).
```

> title: Pod Security Standards

Reviewer: ```suggestion
title: Pod Security Standards (PSS) & Pod Security Admissions (PSA)
```

See my comment further down about the Rancher GUI.

Also:  It's just hard to talk about either resource without mentioning the other.

It's tempting to pull PSACT content into its own page, since how-to create and edit PSACTs feels like its own topic. That leaves the PSS page with the intro, a section on migration, and a section on Rancher namespace exemptions. Thoughts?

Author: > It's tempting to pull PSACT content into its own page, since how-to create and edit PSACTs feels like its own topic. That leaves the PSS page with the intro, a section on migration, and a section on Rancher namespace exemptions. Thoughts?

@martyav I agree, but I'm not sure what are the plans about expanding the docs further.

@jiaqiluo do you plan to address ^ in https://github.com/rancher/rancher-docs/issues/462?

Author: I think that a high-level section about PSS/PSA and then a subpage about PSACT would make sense.

Reviewer 5: hi @macedogm, I agree with your idea.   
@martyav will handle #462 , so I will leave it to him to finalize the structure. 

I think we do not need to hold this PR because of that. We can get this PR merged and use #462  to do the refactor or enhancement.

Author: Just a note, before getting Marty's feedback about merging this PR or not, we still need to merge his suggestion above. I haven't done this yet, because it will close this thread (that is still going on).

Reviewer: I think it's a good idea to merge this PR if there is nothing else to add, then I can create a working branch off of the 2.7.2 release branch, and open a PR for the PSACT how-to. It'll be easier to review the how-to that way.
